scapy/layers/l2.py

```diff
@@ -79,7 +79,7 @@ def getmacbyip(ip, chainCC=0):
     if (tmp[0] & 0xf0) == 0xe0:  # mcast @
         return "01:00:5e:%.2x:%.2x:%.2x" % (tmp[1] & 0x7f, tmp[2], tmp[3])
     iff, _, gw = conf.route.route(ip)
-    if ((iff == conf.loopback_name) or (ip == conf.route.get_if_bcast(iff))):  # noqa: E501
+    if ((iff == conf.loopback_name) or (ip in conf.route.get_if_bcast(iff))):  # noqa: E501
```

Comment:
Maybe since you are changing this, we could remove the `noqa`?
```suggestion
    if (
            iff == conf.loopback_name or
            ip in conf.route.get_if_bcast(iff)
    ):
```

Reply:
Thanks Pierre - that helps me understand what I'd goofed, and I'll so as you suggest shortly (I'm working at the moment):
1) The #noqa: E501 suppresses a 'line too long' warning, so can only come off if the new code doesn't trigger that warning without it?
2) I'll rebase my fork from the master (D'oh!)
3) I'll squash my commits (noob error, sorry)

Reply:
1. Totally! A short line won't trigger the warning. When we added flake8 checks, we decided not no change every line that was too long and simply add a `noqa`. But when we need to change such a line for another reason, we use the opportunity to try to fix it. BTW, that's just a suggestion, you don't have to do that.
2 & 3: perfect! Thanks!

Reply:
I *think* I've now done what's needed but I'm painfully aware this is the first contribution I've made to a busy project and I may have screwed up the rebase / merge operations... If so, any pointers to good Github-for-Dummies guides would be appreciated.

Reply:
Belay that - my regression test is now failing so I've definitely done something wrong.

Reply:
I confess I'm baffled, as the tests work in my local copy with run_scapy. Is this a Python2/3 problem as I'm using Python3? But the run_tests -2 throws no errors for me... help?
```
>>> conf.route.routes = [ 
...:     (0, 0, '172.21.230.1', 'scapy', '172.21.230.10', 1),                # 0.0.0.0         / 0.0.0.0         == 255.255.255.255 
...:     (2130706432, 4278190080, '0.0.0.0', 'lo0', '127.0.0.1', 1),         # 127.0.0.0       / 255.0.0.0       == 127.255.255.255 
...:     (2130706433, 4294967295, '0.0.0.0', 'lo0', '127.0.0.1', 1),         # 127.0.0.1       / 255.255.255.255 == 127.0.0.1 
...:     (2851995648, 4294901760, '0.0.0.0', 'scapy', '172.21.230.10', 1),   # 169.254.0.0     / 255.255.0.0     == 169.254.255.255 
...:     (2887116288, 4294967040, '0.0.0.0', 'scapy', '172.21.230.10', 1),   # 172.21.230.0    / 255.255.255.0   == 172.21.230.255 
...:     (2887116289, 4294967295, '0.0.0.0', 'scapy', '172.21.230.10', 1),   # 172.21.230.1    / 255.255.255.255 == 172.21.230.1 
...:     (3758096384, 4026531840, '0.0.0.0', 'scapy', '172.21.230.10', 1),   # 224.0.0.0       / 240.0.0.0       == 239.255.255.255 
...:     (3758096635, 4294967295, '0.0.0.0', 'scapy', '172.21.230.10', 1),   # 224.0.0.251     / 255.255.255.255 == 224.0.0.251 
...:     (4294967295, 4294967295, '0.0.0.0', 'scapy', '172.21.230.10', 1),   # 255.255.255.255 / 255.255.255.255 == 255.255.255.255 
...:     ] 
...:                                                                                                                                                                               
>>> assert sorted(conf.route.get_if_bcast("scapy")) == sorted(['169.254.255.255', '172.21.230.255', '239.255.255.255']) 
...:                                                                                                                                                                               
>>> conf.route.get_if_bcast("scapy")                                                                                                                                               
['169.254.255.255', '172.21.230.255', '239.255.255.255']
```

```
iMac-2017:scapy michthom$ cd ~/git/scapy
iMac-2017:scapy michthom$ cd test
iMac-2017:test michthom$ ./run_tests -2
### Non-root mode ###
### Booting scapy...
### Starting tests...
### Writing output...
iMac-2017:test michthom$ echo $?
0
iMac-2017:test michthom$ ./run_tests -3
### Non-root mode ###
### Booting scapy...
### Starting tests...
### Writing output...
iMac-2017:test michthom$ echo $?
0
```

Reply:
That's the purpose of these tests: prevent us from merging something that would work in an environment and not in another one.

Reply:
If it may help, the broken tests run under Windows.

Reply:
I'm a BIG fan of automated testing. Ah. I think my problem was I took the comment by @guedou above too literally and used a made-up interface called 'scapy' - whereas I need to use the one returned by the ```get_dummy_interface() method``` which has the GUID populated? But then in my "fake" routes I'd also need to populate the interface in each of those with the same object?

Reply:
Argh. I've tied myself in a knot and don't know how to get free. I've resync'd to the upstream HEAD, pulled that and pushed to my forked repo. I'm trying to squash my latest commit but I'm now seeing multiple instances of my commits (pre-squash and post) and other's changes now that can't be applied.

Reply:
> If it may help, the broken tests run under Windows.

Sadly not. I've put what I believe to be correct code in place, and it's *still* failing in the Windows build, complaining that str has no .guid attribute. Which makes no sense... unless the WINDOWS flag is not getting set before the tests are run in the automated environment, so the get_dummy_device() returns "dummy0" as a string??

My reading of the code says that if it's Windows, then get_dummy_interface() returns a NetworkInterface populated with a .guid attribute. But the error says it's just a string...

Reply:
OK, I've got a windows VM set up with Python 3.8 and Scapy git-archive.dev02653d5053.
It behaves as expected with the unmodified Scapy from upstream.
Now to find out if my modifications genuinely make it barf (and if so why?)
```
>>> (interface, localIP, defaultGateway) = conf.route.route("0.0.0.0")
>>> interface
<NetworkInterface [Intel(R) 82574L Gigabit Network Connection] {26281F51-C4E8-47A8-9B1B-B796D9DACEC5}>
>>> interface.guid
'{26281F51-C4E8-47A8-9B1B-B796D9DACEC5}'
>>> (interface, localIP, defaultGateway) = conf.route.route("0.0.0.0")
>>> for route in conf.route.routes:
...     (net, msk, gw, iface, addr, metric) = route
...     print(ltoa(net), ltoa(msk), gw, iface, addr, metric)
...
0.0.0.0 0.0.0.0 172.16.198.2 <NetworkInterface [Intel(R) 82574L Gigabit Network Connection] {26281F51-C4E8-47A8-9B1B-B796D9DACEC5}> 172.16.198.130 25
127.0.0.0 255.0.0.0 0.0.0.0 <NetworkInterface [Software Loopback Interface 1] \Device\NPF_Loopback> 127.0.0.1 331
127.0.0.1 255.255.255.255 0.0.0.0 <NetworkInterface [Software Loopback Interface 1] \Device\NPF_Loopback> 127.0.0.1 331
127.255.255.255 255.255.255.255 0.0.0.0 <NetworkInterface [Software Loopback Interface 1] \Device\NPF_Loopback> 127.0.0.1 331
172.16.198.0 255.255.255.0 0.0.0.0 <NetworkInterface [Intel(R) 82574L Gigabit Network Connection] {26281F51-C4E8-47A8-9B1B-B796D9DACEC5}> 172.16.198.130 281
172.16.198.130 255.255.255.255 0.0.0.0 <NetworkInterface [Intel(R) 82574L Gigabit Network Connection] {26281F51-C4E8-47A8-9B1B-B796D9DACEC5}> 172.16.198.130 281
172.16.198.255 255.255.255.255 0.0.0.0 <NetworkInterface [Intel(R) 82574L Gigabit Network Connection] {26281F51-C4E8-47A8-9B1B-B796D9DACEC5}> 172.16.198.130 281
224.0.0.0 240.0.0.0 0.0.0.0 <NetworkInterface [Software Loopback Interface 1] \Device\NPF_Loopback> 127.0.0.1 331
224.0.0.0 240.0.0.0 0.0.0.0 <NetworkInterface [Intel(R) 82574L Gigabit Network Connection] {26281F51-C4E8-47A8-9B1B-B796D9DACEC5}> 172.16.198.130 281
224.0.0.0 240.0.0.0 0.0.0.0 <NetworkInterface [Bluetooth Device (Personal Area Network) #2] {9C31EA74-A233-4C25-A7AF-C440DDADD100}> 169.254.82.33 321
255.255.255.255 255.255.255.255 0.0.0.0 <NetworkInterface [Software Loopback Interface 1] \Device\NPF_Loopback> 127.0.0.1 331
255.255.255.255 255.255.255.255 0.0.0.0 <NetworkInterface [Intel(R) 82574L Gigabit Network Connection] {26281F51-C4E8-47A8-9B1B-B796D9DACEC5}> 172.16.198.130 281
255.255.255.255 255.255.255.255 0.0.0.0 <NetworkInterface [Bluetooth Device (Personal Area Network) #2] {9C31EA74-A233-4C25-A7AF-C440DDADD100}> 169.254.82.33 321```

Reply:
I'm back to baffled. As @p-l- mentioned, the tests in my new code pass, at least on my Windows VM. So I don't understand why the automated tests are failing. @guedou I'm more than happy to try to fix but I'm going to need guidance on what might be causing this?